arroyo-worker/src/operators/windows.rs

```diff
@@ -135,51 +179,188 @@ impl<K: Key, T: Data, OutT: Data, W: TimeWindowAssigner<K, T>> KeyedWindowFunc<K
+            .unwrap_or(SystemTime::UNIX_EPOCH);
+
+        if watermark >= record.timestamp {
+            // drop late data
+            return;
+        }
+
+        let mut key = record.key.as_ref().unwrap().clone();
+        let value = record.value.clone();
+        let timestamp = record.timestamp;
+
+        let mut windows: Option<Vec<Window>> = {
+            let t: KeyedState<'_, K, Vec<Window>, _> = ctx.state.get_key_state('s').await;
+            t.get(&key).map(|t| t.iter().map(|w| *w).collect())
+        };
+
+        let (remove, add) = if let Some(windows) = windows.as_mut() {
+            if let Some((i, window)) = windows
+                .iter()
+                .enumerate()
+                .find(|(_, w)| w.contains(timestamp))
+                .map(|(i, w)| (i, *w))
+            {
+                // there's an existing window this record falls into
+                let our_end = timestamp + self.gap_size;
+                if our_end > window.end {
+                    // we're extending an existing window
+                    windows.remove(i);
+                    let new = (window.start..our_end).into();
+                    (Some(window.end), Self::handle_new_window(windows, new))
+                } else {
+                    (None, None)
                 }
+                // otherwise the window is unchanged, we don't need to do anything
+            } else {
+                // there's no existing window, we need to add one
+                (
+                    None,
+                    Self::handle_new_window(windows, Window::session(timestamp, self.gap_size)),
+                )
             }
+        } else {
+            // no existing window, create one
+            let window = Window::session(timestamp, self.gap_size);
+            windows = Some(vec![]);
+            let t = Self::handle_new_window(windows.as_mut().unwrap(), window);
+            (None, t)
+        };
+
+        if let Some(remove) = remove {
+            let _: Option<Window> = ctx.cancel_timer(&mut key, remove).await;
+        }
+        if let Some(add) = add {
+            // we use UNIX_EPOCH as the start of our windows to avoid having to update them when we extend
+            // the beginning; this works because windows are always handled in order
+            ctx.schedule_timer(&mut key, add, Window::new(SystemTime::UNIX_EPOCH, add))
+                .await;
         }
 
-        // clear everything before our start time (we're guaranteed that timers execute in order,
-        // so with fixed-width windows there won't be any earlier data)
-        let next = self.assigner.next(window);
+        if add.is_some() || remove.is_some() {
+            let key = key.clone();
+            ctx.state
+                .get_key_state('s')
+                .await
+                // I have no idea if this timestamp is correct -- this datastructure does not make sense to me
+                .insert(timestamp, key, windows.unwrap())
```

Comment:
I think this is wrong. Given that events can arrive out of order, imagine you have a 24 hour window in the future, with gaps of 30 minutes, it all fills in at 00:00, but the last event you process is at 00:10. This'd get written with that 00:10 timestamp. Then, if you restarted at 1:00, it would expire the record of your window. This should be the max of all timestamps you've seen for the key, I believe.

Reply:
You can compute this by maxing over the windows, or storing it along with the vec of Windows.

Reply:
Good point, this works for existing windows because they're fixed width, so the watermark + the safe_retention_micros is enough to ensure we don't delete data we might need in the future.

However I'm not sure your approach would work either, since at insertion time we can't know how long we need to store it—every time the window gets extended we'd have to rewrite all of the events for that key.

Instead what do you think about just opting out of the optimized delete behavior and relying on the `clear_time_range` calls in handle_timer?

Reply:
My comments were for the 's' table, which gets rewritten every time we change the windows. For the window state table 's', you are correct that we can't use the expiration, so should be relying on clear_time_range.

Reply:
From offline conversation:

* For the window table 'w' we can get the correct behavior by using the max endtime of all of the windows, which ensures that we do not clear up the windows vec until all windows are finished
* The state table 's' is more problematic; since we can't know at insertion time when the window that contains that record will close (since it can be arbitrarily extended) we can't use any watermark-based cleanup—currently the only type of deletion currently supported by the state backend. For now, we will introduce a max session window size of 24 hours; any sessions longer than that will get broken up. All state will be retained for 24 hours then deleted. As a follow on, we will add support for tombstone-based deletion and compaction in the s3 state backend.